src/compiler/watchPublic.ts

```diff
@@ -823,6 +830,24 @@ namespace ts {
                 watchFilePath(missingFilePath, missingFilePath, onMissingFileChange, PollingInterval.Medium, watchOptions, WatchType.MissingFile);
         }
 
+        function watchPackageJsonLookupPath(packageJsonPath: Path) {
+            // If the package.json is pulled into the compilation itself (eg, via json imports), don't add a second watcher here
+            return sourceFilesCache.has(packageJsonPath) ?
+                noopFileWatcher :
+                watchFilePath(packageJsonPath, packageJsonPath, onPackageJsonChange, PollingInterval.High, watchOptions, WatchType.PackageJson);
+        }
+
+        function onPackageJsonChange(fileName: string, eventKind: FileWatcherEventKind, path: Path) {
+            updateCachedSystemWithFile(fileName, path, eventKind);
+
+            // package.json changes invalidate module resolution and can change the set of loaded files
+            // so if we witness a change to one, we have to do a full reload
+            reloadLevel = ConfigFileProgramReloadLevel.Full;

```

Comment:
This seems incorrect.. Full Reload means read config file again to calculate file names.. Why do we need to do that for package.json changes. 
Also changesAffectResolution seems very unoptimized way of saying resolve all modules again. Need smarter way to invalidate resolutions.. Esp when someone is not using the node12+ stuff.. looks like every npm install will be problematic? (it many times spans across multiple updates if its many packages)

Reply:
A package.json change can change what arbitrary module specifiers in the program resolve to (when main or exports change) and which imports are valid in arbitrary files (when type changes), so, yeah, a package.json change can invalidate the whole program.

Reply:
And do note, only the exports and type fields are new - we failed to pickup main and typesVersions changes and invalidations from package.json files entirely before. A fix for that actually landed _before_ any of the node12 stuff.

Reply:
But invalidating whole program is different from reading config file and getting initial set which is what  ConfigFileProgramReloadLevel.Full means

Reply:
`package.json` main is about as far from "rare" as you can get - every package has one. :P

You would be right that we probably don't need to watch them if moduleResolution isn't node/node12/nodenext though. Not that many people use TS without one of those set, since `module: commonjs` (and thus `moduleResolution: node`) is the norm. Except I believe `@types` resolution unconditionally uses `node`-style resolution anyway; so I think we actually need to watch the `package.json` documents in those folders unconditionally...